components/common/src/templating/config.rs

```diff
@@ -587,35 +573,109 @@ fn is_toml_value_a_table(key: &str, table: &toml::value::Table) -> bool {
     }
 }
 
-#[cfg(not(windows))]
-fn set_permissions<T: AsRef<Path>>(path: T, pkg: &Pkg) -> hcore::error::Result<()> {
+#[cfg(unix)]
+fn set_permissions(path: &Path, user: &str, group: &str) -> hcore::error::Result<()> {
     use crate::hcore::os::users;
     use crate::hcore::util::posix_perm;
 
     if users::can_run_services_as_svc_user() {
-        posix_perm::set_owner(path.as_ref(), &pkg.svc_user, &pkg.svc_group)?;
+        posix_perm::set_owner(path, &user, &group)?;
     }
-    posix_perm::set_permissions(path.as_ref(), CONFIG_PERMISSIONS)
+
+    let permissions = if path.is_dir() {
+        CONFIG_DIR_PERMISSIONS
+    } else {
+        CONFIG_PERMISSIONS
+    };
+    posix_perm::set_permissions(&path, permissions)
 }
 
 #[cfg(windows)]
-fn set_permissions<T: AsRef<Path>>(path: T, _pkg: &Pkg) -> hcore::error::Result<()> {
+fn set_permissions(path: &Path, _user: &str, _group: &str) -> hcore::error::Result<()> {
     use crate::hcore::util::win_perm;
 
-    win_perm::harden_path(path.as_ref())
+    win_perm::harden_path(&path)
+}
+
+/// Recursively walk the configuration directory and subdirectories to
+/// construct the list of template files
+///
+/// `dir` should be a directory that exists.
+fn load_templates(
+    dir: &Path,
+    context: &Path,
+    mut template: TemplateRenderer,
+) -> Result<TemplateRenderer> {
+    for entry in std::fs::read_dir(dir)?.filter_map(|entry| entry.ok()) {
+        // We're storing the pathname relative to the input config directory
+        // as the identifier for the template
+        let relative_path = context.join(&entry.file_name());
+        match entry.file_type() {
+            Ok(file_type) if file_type.is_file() => {
+                // JW TODO: This error needs improvement. TemplateFileError is too generic.
+                template
+                    .register_template_file(&relative_path.to_string_lossy(), &entry.path())
+                    .map_err(Error::TemplateFileError)?;
+            }
+            Ok(file_type) if file_type.is_dir() => {
+                template = load_templates(&entry.path(), &relative_path, template)?
+            }
+            Ok(file_type) => trace!("Skipping non file/directory entry: {:?}", file_type),
+            Err(e) => {
+                debug!("Failed to get file metadata for {:?} : {}", entry, e);
+                continue;
```

Comment:
I think this `continue` is superfluous. Not a big deal, but eliminating it might obviate the braces for the `match` arm and save a couple lines.